packages/router/src/adapters/contract/contract.ts

```diff
@@ -130,53 +89,100 @@ export const cancelAndFullfillSanitationCheck = async (
     data: encodeVariantTransactionData,
   });
 
-  if (expectedVariantDigest === variantTransactionDigest) {
-    // All is good, no issues
-    return;
-  }
-
-  // transaction should be prepared before fulfill
-  if (variantTransactionDigest === HashZero) {
-    const error = new Error("Transaction isn't prepared yet");
-    logger.error(
-      "FAILED cancelAndFullfillSanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT: Transaction isn't prepared yet",
-      requestContext,
-      methodContext,
-      jsonifyError(error),
-      { transactionData, chainId },
-    );
-    throw error;
+  if (functionCall === "prepare") {
+    // variantTransactionDigest exist then transaction is already prepared
+    if (variantTransactionDigest !== HashZero) {
+      const error = new Error("Transaction is already prepared");
+      logger.error(
+        `FAILED ${functionCall} sanitationCheck THIS SHOULD NOT HAPPEN, FIGURE THIS OUT`,
+        requestContext,
+        methodContext,
+        jsonifyError(error),
+        { transactionData, chainId },
+      );
+      throw error;
+    }
+  } else {
+    const expectedVariantDigest = getVariantTransactionDigest({
+      amount: transactionData.amount,
+      expiry: transactionData.expiry,
+      preparedBlockNumber: transactionData.preparedBlockNumber,
+    });
+
+    if (expectedVariantDigest === variantTransactionDigest) {
+      // All is good, no issues
+      return;
+    }
+
+    // transaction should be prepared before fulfill
```

Comment:
this could also be cancel in this block, please clarify in this comment
`// transaction should be prepared before fulfill or cancel`